File: challenges/c04_summarizer/solution_summarizer.py
"""
CLI for text summarization using Click.

Allows summarizing a text file using Hugging Face models with options for:
- Summary type: short, medium, bullet
- Model override
- API token override
"""

import click
from summarizer import summarize

from common.logging_config import get_logger

logger = get_logger(__name__)


@click.group()
def cli() -> None:
    """Summarizer CLI group using Click."""
    pass


@cli.command(name="summarize")
@click.option(
    "--input",
    "-i",
    "input_file",
    type=click.Path(exists=False),
    required=False,
    help="Path to the article file. Defaults to article.txt if not provided.",
)
@click.option(
    "--type",
    "-t",
    "summary_type",
    required=True,
    type=click.Choice(["short", "medium", "bullet"]),
    help="Type of summary to generate.",
)
@click.option(
    "--model",
    "-m",
    help="Hugging Face model override (optional).",
)
@click.option(
    "--api-token",
    help="Hugging Face API token override (optional).",
)
def cli_summarize(input_file: str, summary_type: str, model: str, api_token: str) -> None:
    """
    Summarizes a text file using the Hugging Face model.

    Args:
        input_file (str): Path to input text file. Defaults to article.txt if None.
        summary_type (str): 'short', 'medium', or 'bullet'.
        model (str): Optional Hugging Face model override.
        api_token (str): Optional Hugging Face API token override.
    """
    if input_file:
        logger.info(f"[CLI] Using provided file: {input_file}")
        file_path = input_file
    else:
        logger.info("[CLI] No input provided → defaulting to article.txt")
        file_path = None  # summarizer handles default file

    # Run summarization
    result = summarize(
        text_path=file_path,
        summary_type=summary_type,
        cli_model=model,
        cli_token=api_token,
    )

    # Output result
    click.echo("\n=== SUMMARY RESULT ===\n")
    click.echo(result)


if __name__ == "__main__":
    cli()

File: challenges/c04_summarizer/solution_summarizer.go
package main

import (
	"fmt"
	"os"

	"DataSecChallenge/challenges/c04_summarizer/summarizer" // Import our summarizer package

	"github.com/spf13/cobra"

	"github.com/joho/godotenv" // Importar godotenv
)

// GoLang version used: 1.25.5

const envFile = ".env"

var (
	inputFile   string
	summaryType string
	model       string
	apiToken    string
)

var rootCmd = &cobra.Command{
	Use:   "summarize",
	Short: "CLI for text summarization using Hugging Face GenAI",
	Long: `A command-line application that summarizes the contents of a text file using a free, public GenAI API (HuggingFace Inference API).
Requirements:
The CLI must be written in Go.
The CLI must accept:
  --input or a positional argument: path to the text file to summarize.
  --type or -t : summary type, one of short, medium, or bullet.
The CLI must call a free, public GenAI API for summarization (e.g., HuggingFace Inference API).
The prompt sent to the API should be engineered to match the summary type:
  short : a concise summary (1-2 sentences)
  medium : a paragraph summary
  bullet : a list of bullet points
The CLI should output the summary to stdout.
The CLI should handle API errors gracefully and print user-friendly messages.
Document the Go version used in your code comments.`,
	Run: func(cmd *cobra.Command, args []string) {
		if summaryType == "" {
			fmt.Println("Error: --type is required (short, medium, bullet)")
			os.Exit(1)
		}

		// Handle default input file if not provided as flag, check positional args
		if inputFile == "" && len(args) > 0 {
			inputFile = args[0]
		}

		if inputFile == "" {
			fmt.Println("[CLI] No input provided → defaulting to article.txt")
			// Note: The LoadArticle function handles the full path for the default article.
		}

		// Load article content
		articleContent, err := summarizer.LoadArticle(inputFile)
		if err != nil {
			fmt.Printf("Error loading article: %v\n", err)
			os.Exit(1)
		}

		finalToken := apiToken
		if finalToken == "" {
			finalToken = os.Getenv("HF_API_TOKEN")
		}

		// Perform summarization
		summaryResult, err := summarizer.SummarizeText(articleContent, summaryType, model, finalToken, nil)
		if err != nil {
			fmt.Printf("Error during summarization: %v\n", err)
			os.Exit(1)
		}

		// Output result
		fmt.Println("\n=== SUMMARY RESULT ===")
		fmt.Println(summaryResult)
	},
}

func init() {
	rootCmd.PersistentFlags().StringVarP(&inputFile, "input", "i", "", "Path to the article file.")
	rootCmd.PersistentFlags().StringVarP(&summaryType, "type", "t", "", "Type of summary to generate (short, medium, bullet).")
	rootCmd.PersistentFlags().StringVarP(&model, "model", "m", "", "Hugging Face model override (optional).")
	rootCmd.PersistentFlags().StringVar(&apiToken, "api-token", "", "Hugging Face API token override (optional).")
}

func main() {
	/* Como este programa es un CLI, se ejecuta desde el directorio raiz
	 * del proyecto, se carga el archivo .env desde el directorio raiz.
	 * este es el valor si se corre desde dentro del directorio "c04-summarizer"
	 *antes _ = godotenv.Load("../../.env")
	 */
	// Intentamos cargar el .env
	err := godotenv.Load(envFile)

	// Si hay un error, verificamos si es porque el archivo no existe
	if err != nil {
		// Imprimimos un mensaje amigable pero informativo
		fmt.Println("💡 [INFO] No se encontró el archivo .env. Se usarán las variables de entorno del sistema.")
		fmt.Println("   (Si necesitas configurar HF_API_TOKEN, crea un archivo .env en la raíz")
		fmt.Println("   usa .env.exmaple como referencia)")
		fmt.Println("-----------------------------------------------------------------------")
	}

	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}

	if err := rootCmd.Execute(); err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}
